File: lib/wasm/events.ts
/**
 * Decodes the wire format `h3c_drain_events`/`qc_drain_events` return: one
 * JSON array using the same field names as the native NAPI event contract
 * (`eventType`, `streamId`, `headers`, `fin`, `meta`, `metrics` —
 * `crates/http3-wasm/src/events.rs`), except binary payloads, which arrive
 * as `dataOff`/`dataLen` — offsets into the wasm module's own linear
 * memory, valid only until the next ABI call on the handle
 * (`crates/http3-wasm/src/lib.rs`'s buffer-lifetime rule).
 *
 * This module deliberately does NOT import `NativeEvent` from
 * `lib/event-loop.ts` — the `lib/wasm/**` ESLint zone forbids importing
 * that file (docs/WASM_CLIENT_PLAN.md §6.6). {@link WasmEvent} below is a
 * structurally *almost* identical shape, kept in sync by hand, with one
 * deliberate difference (Phase 5, docs/WASM_CLIENT_PLAN.md §9): binary
 * fields are `Uint8Array`, not Node's `Buffer` — this file must stay
 * compilable with no `@types/node` in scope (a future workerd host has
 * none). Because of that one difference, a `(events: WasmEvent[]) => void`
 * callback is **not** directly assignable to a real
 * `(events: NativeEvent[]) => void` callback (e.g. `session._dispatchEvents`)
 * the way it was pre-Phase-5 — `lib/client.ts`/`lib/quic-client.ts` (outside
 * `lib/wasm/`, Node-only, free to import both types) now do one small,
 * explicit conversion step (wrapping every `Uint8Array` payload in a real
 * `Buffer.from(...)`) right at that boundary, so every existing,
 * documented, Buffer-typed public API keeps receiving exactly what it did
 * before this phase — see `lib/wasm-event-bridge.ts`.
 */

import type { Http3WasmCore } from './core-loader.js';

/** Mirrors `NativeEvent` in `lib/event-loop.ts` field-for-field, except binary payloads are `Uint8Array` — see the module doc comment. */
export interface WasmEvent {
  eventType: number;
  connHandle: number;
  streamId: number;
  headers?: Array<{ name: string; value: string }>;
  data?: Uint8Array;
  fin?: boolean;
  meta?: {
    errorCode?: number;
    errorReason?: string;
    errorCategory?: string;
    remoteAddr?: string;
    remotePort?: number;
    serverName?: string;
    reasonCode?: string;
    runtimeDriver?: string;
    runtimeMode?: string;
    requestedRuntimeMode?: string;
    fallbackOccurred?: boolean;
    errno?: number;
    syscall?: string;
    peerCertificatePresented?: boolean;
    /** Never actually populated on either runtime today (WASM_CLIENT_PLAN.md §12 decision log) — kept only for shape parity. */
    peerCertificateChain?: Uint8Array[];
    durationMs?: number;
  };
  metrics?: {
    packetsIn: number;
    packetsOut: number;
    bytesIn: number;
    bytesOut: number;
    handshakeTimeMs: number;
    rttMs: number;
    cwnd: number;
    datagramQueueDepth: number;
  };
}

/** Raw per-event JSON shape `serde_json` produces (`crates/http3-wasm/src/events.rs`). */
interface RawWasmEventJson {
  eventType: number;
  streamId: number;
  headers?: Array<{ name: string; value: string }>;
  fin?: boolean;
  dataOff?: number;
  dataLen?: number;
  meta?: WasmEvent['meta'];
  metrics?: WasmEvent['metrics'];
}

/**
 * Decode a `drain_events` JSON array and copy every `dataOff`/`dataLen`
 * payload out of linear memory into a fresh `Uint8Array` **synchronously**,
 * right now — before the caller makes any further ABI call on this
 * handle. `connHandle` is stamped onto every event because the wire JSON
 * omits it: the native NAPI event batch carries a `connHandle` per event
 * for the multi-connection worker-thread case, but a wasm session is
 * always exactly one connection, so the caller (which already owns the
 * handle number) supplies it once here instead of threading it through
 * the Rust->JSON->JS round trip.
 */
export function decodeEventBatch(core: Http3WasmCore, json: string, connHandle: number): WasmEvent[] {
  const raw = JSON.parse(json) as RawWasmEventJson[];
  const events: WasmEvent[] = [];
  for (const e of raw) {
    const event: WasmEvent = {
      eventType: e.eventType,
      connHandle,
      streamId: e.streamId,
    };
    if (e.headers) event.headers = e.headers;
    if (e.fin !== undefined) event.fin = e.fin;
    if (e.meta) event.meta = e.meta;
    if (e.metrics) event.metrics = e.metrics;
    if (typeof e.dataOff === 'number' && typeof e.dataLen === 'number' && e.dataLen > 0) {
      event.data = core.copyOut(e.dataOff, e.dataLen);
    }
    events.push(event);
  }
  return events;
}

/**
 * Drain a `*_take_keylog(handle, outPtrPtr)` export (`h3c_take_keylog` or
 * `qc_take_keylog` — pass the bound export function directly; both share
 * this exact shape) and copy out any accumulated NSS-format keylog lines.
 * Returns `null` when nothing has accumulated since the last call.
 */
export function drainKeylog(
  core: Http3WasmCore,
  takeKeylogFn: (handle: number, outPtrPtr: number) => bigint,
  handle: number,
  outPtrPtr: number,
): Uint8Array | null {
  const len = Number(takeKeylogFn(handle, outPtrPtr));
  if (len <= 0) return null;
  return core.readOutPtrResult(outPtrPtr, len);
}

File: crates/http3-wasm/src/events.rs
//! Serializes `JsH3Event` batches into the wire format `h3c_drain_events` /
//! `qc_drain_events` return: one JSON array using the exact TS field names
//! (`eventType`, `streamId`, `headers`, `fin`, `meta`, `metrics`), with
//! `data` payloads replaced by `dataOff`/`dataLen` — absolute offsets into
//! this module's own linear memory (the same address space `h3c_rx_buffer`
//! etc. already hand out pointers into), valid only until the next call on
//! the handle (see the crate-level doc comment).

use http3::wasm_exports::{JsEventMeta, JsH3Event, JsSessionMetrics};
use serde_json::{Map, Value};

/// Serialize `events` into `json_scratch` (cleared and overwritten) and
/// append every event's `data` payload into `data_scratch` (cleared first,
/// then grown once per event) so callers can compute a single stable base
/// pointer after all pushes are done — `Vec::as_ptr()` is only meaningful
/// once nothing will trigger a reallocation.
pub(crate) fn serialize_events(
    events: &[JsH3Event],
    json_scratch: &mut Vec<u8>,
    data_scratch: &mut Vec<u8>,
) {
    data_scratch.clear();
    let mut values: Vec<Value> = Vec::with_capacity(events.len());
    for event in events {
        values.push(event_to_json_relative(event, data_scratch));
    }

    // `data_scratch` is now fully built and will not reallocate again
    // before the caller reads it — safe to take one absolute base pointer
    // and bake it into every event's relative offset.
    let base = data_scratch.as_ptr() as u64;
    for value in &mut values {
        let Value::Object(obj) = value else { continue };
        let Some(off) = obj.get("dataOff").and_then(Value::as_u64) else {
            continue;
        };
        obj.insert("dataOff".to_string(), Value::from(off + base));
    }

    json_scratch.clear();
    // `Vec<u8>` implements `std::io::Write` — no separate string buffer.
    let _ = serde_json::to_writer(&mut *json_scratch, &Value::Array(values));
}

fn event_to_json_relative(event: &JsH3Event, data_scratch: &mut Vec<u8>) -> Value {
    let mut obj = Map::new();
    obj.insert("eventType".to_string(), Value::from(event.event_type));
    obj.insert("streamId".to_string(), Value::from(event.stream_id));

    if let Some(fin) = event.fin {
        obj.insert("fin".to_string(), Value::from(fin));
    }

    if let Some(headers) = &event.headers {
        let arr: Vec<Value> = headers
            .iter()
            .map(|h| {
                let mut hm = Map::new();
                hm.insert("name".to_string(), Value::from(h.name.clone()));
                hm.insert("value".to_string(), Value::from(h.value.clone()));
                Value::Object(hm)
            })
            .collect();
        obj.insert("headers".to_string(), Value::Array(arr));
    }

    if let Some(data) = &event.data {
        // Relative offset for now; `serialize_events` rewrites this to an
        // absolute linear-memory address once `data_scratch` is final.
        let off = data_scratch.len() as u64;
        data_scratch.extend_from_slice(data);
        obj.insert("dataOff".to_string(), Value::from(off));
        obj.insert("dataLen".to_string(), Value::from(data.len() as u64));
    }

    if let Some(meta) = &event.meta {
        obj.insert("meta".to_string(), meta_to_json(meta));
    }

    if let Some(metrics) = &event.metrics {
        obj.insert("metrics".to_string(), metrics_to_json(metrics));
    }

    Value::Object(obj)
}

fn meta_to_json(meta: &JsEventMeta) -> Value {
    let mut obj = Map::new();
    if let Some(v) = meta.error_code {
        obj.insert("errorCode".to_string(), Value::from(v));
    }
    if let Some(v) = &meta.error_reason {
        obj.insert("errorReason".to_string(), Value::from(v.clone()));
    }
    if let Some(v) = &meta.error_category {
        obj.insert("errorCategory".to_string(), Value::from(v.clone()));
    }
    if let Some(v) = &meta.remote_addr {
        obj.insert("remoteAddr".to_string(), Value::from(v.clone()));
    }
    if let Some(v) = meta.remote_port {
        obj.insert("remotePort".to_string(), Value::from(v));
    }
    if let Some(v) = &meta.server_name {
        obj.insert("serverName".to_string(), Value::from(v.clone()));
    }
    if let Some(v) = &meta.reason_code {
        obj.insert("reasonCode".to_string(), Value::from(v.clone()));
    }
    if let Some(v) = &meta.runtime_driver {
        obj.insert("runtimeDriver".to_string(), Value::from(v.clone()));
    }
    if let Some(v) = &meta.runtime_mode {
        obj.insert("runtimeMode".to_string(), Value::from(v.clone()));
    }
    if let Some(v) = &meta.requested_runtime_mode {
        obj.insert("requestedRuntimeMode".to_string(), Value::from(v.clone()));
    }
    if let Some(v) = meta.fallback_occurred {
        obj.insert("fallbackOccurred".to_string(), Value::from(v));
    }
    if let Some(v) = meta.errno {
        obj.insert("errno".to_string(), Value::from(v));
    }
    if let Some(v) = &meta.syscall {
        obj.insert("syscall".to_string(), Value::from(v.clone()));
    }
    if let Some(v) = meta.peer_certificate_presented {
        obj.insert("peerCertificatePresented".to_string(), Value::from(v));
    }
    // `peerCertificateChain` deliberately omitted: the client path never
    // populates it today (native or wasm — see docs/WASM_CLIENT_PLAN.md
    // decision log "peer_certificate_chain on client HANDSHAKE_COMPLETE —
    // deferred"), and encoding a `Vec<Vec<u8>>` through the scratch-buffer
    // scheme isn't worth the complexity until something actually sets it.
    if let Some(v) = meta.duration_ms {
        obj.insert("durationMs".to_string(), Value::from(v));
    }
    Value::Object(obj)
}

/// Standalone session-metrics JSON string, for `h3c_session_metrics` /
/// `qc_session_metrics` (outside the per-event batch context).
pub(crate) fn metrics_only_json(metrics: &JsSessionMetrics) -> String {
    metrics_to_json(metrics).to_string()
}

/// 8-field TS shape — omits native's extra `pmtu` field (§6.4: "Metrics:
/// 8-field TS shape (omit native's extra pmtu field)").
fn metrics_to_json(metrics: &JsSessionMetrics) -> Value {
    let mut obj = Map::new();
    obj.insert("packetsIn".to_string(), Value::from(metrics.packets_in));
    obj.insert("packetsOut".to_string(), Value::from(metrics.packets_out));
    obj.insert("bytesIn".to_string(), Value::from(metrics.bytes_in));
    obj.insert("bytesOut".to_string(), Value::from(metrics.bytes_out));
    obj.insert(
        "handshakeTimeMs".to_string(),
        Value::from(metrics.handshake_time_ms),
    );
    obj.insert("rttMs".to_string(), Value::from(metrics.rtt_ms));
    obj.insert("cwnd".to_string(), Value::from(metrics.cwnd));
    obj.insert(
        "datagramQueueDepth".to_string(),
        Value::from(metrics.datagram_queue_depth),
    );
    Value::Object(obj)
}

#[cfg(test)]
mod tests {
    use super::serialize_events;
    use http3::wasm_exports::JsH3Event;

    #[test]
    fn empty_batch_serializes_to_empty_array() {
        let mut json_scratch = Vec::new();
        let mut data_scratch = Vec::new();
        serialize_events(&[], &mut json_scratch, &mut data_scratch);
        assert_eq!(json_scratch, b"[]");
    }

    #[test]
    fn data_event_gets_absolute_offset_into_data_scratch() {
        let events = vec![JsH3Event::data(1, 4, vec![1, 2, 3, 4], false, ())];
        let mut json_scratch = Vec::new();
        let mut data_scratch = Vec::new();
        serialize_events(&events, &mut json_scratch, &mut data_scratch);

        let parsed: serde_json::Value = serde_json::from_slice(&json_scratch).unwrap();
        let arr = parsed.as_array().unwrap();
        assert_eq!(arr.len(), 1);
        let obj = &arr[0];
        assert_eq!(obj["eventType"], 4);
        assert_eq!(obj["streamId"], 4);
        assert_eq!(obj["dataLen"], 4);
        let off = obj["dataOff"].as_u64().unwrap();
        let base = data_scratch.as_ptr() as u64;
        assert_eq!(off, base);
        assert_eq!(&data_scratch[..4], &[1, 2, 3, 4]);
    }

    #[test]
    fn multiple_data_events_get_distinct_non_overlapping_offsets() {
        let events = vec![
            JsH3Event::data(1, 4, vec![9, 9], false, ()),
            JsH3Event::data(1, 8, vec![1, 2, 3], true, ()),
        ];
        let mut json_scratch = Vec::new();
        let mut data_scratch = Vec::new();
        serialize_events(&events, &mut json_scratch, &mut data_scratch);
        let parsed: serde_json::Value = serde_json::from_slice(&json_scratch).unwrap();
        let arr = parsed.as_array().unwrap();
        let base = data_scratch.as_ptr() as u64;
        let off0 = arr[0]["dataOff"].as_u64().unwrap();
        let off1 = arr[1]["dataOff"].as_u64().unwrap();
        assert_eq!(off0, base);
        assert_eq!(off1, base + 2);
        assert_eq!(arr[1]["fin"], true);
    }
}
